lib/src/Pos/Logic/Types.hs

```diff
@@ -0,0 +1,204 @@
+      -- Included here because that function is quite complicated; it's not
+      -- clear whether it can be expressed simply in terms of getBlockHeader.:q
+    , getBlockHeaders    :: NonEmpty HeaderHash -> Maybe HeaderHash -> m (Either GetBlockHeadersError (NewestFirst NE BlockHeader))
+      -- Inspired by 'getHeadersFromToIncl', which is apparently distinct from
+      -- 'getHeadersFromManyTo' (getBlockHeaders without the tick above).
+      -- FIXME we must unify these.
+      -- May want to think about giving a streaming-IO interface (pipes, conduit
+      -- or similar).
+    , getBlockHeaders'   :: HeaderHash -> HeaderHash -> m (Either GetBlockHeadersError (Maybe (OldestFirst NE HeaderHash)))
+      -- Get the current tip of chain.
+      -- It's not in Maybe, as getBlock is, because really there should always
+      -- be a tip, whereas trying to get a block that isn't in the database is
+      -- normal.
+    , getTip             :: m (Either GetTipError Block)
+
+      -- | Get state of last adopted BlockVersion. Related to update system.
+    , getAdoptedBVData   :: m BlockVersionData
+
+      -- Give a block header to the logic layer.
+      -- NodeId is needed for first iteration, but will be removed later.
+    , postBlockHeader    :: BlockHeader -> NodeId -> m ()
+
+      -- Tx, update, ssc... 
+      -- Common pattern is:
+      --   - What to do with it when we receive it (key and data).
+      --   - How to get it when it's requested (key).
+      --   - Whether to request it when it's announced (key).
+      --   - How to derive the key from the data.
+      --     Needed in case we get data, so that we can use the key to
+      --     announce it.
+      -- For the first iteration we'll keep changes to this inv/req/data
+      -- system minimal, so the logic layer must define how to do all of
+      -- these things for every relayed piece of data.
+      -- See comment on the 'KeyVal' type.
+    , postTx            :: KeyVal (Tagged TxMsgContents TxId) TxMsgContents m
+    , postUpdate        :: KeyVal (Tagged (UpdateProposal, [UpdateVote]) UpId) (UpdateProposal, [UpdateVote]) m
+    , postVote          :: KeyVal (Tagged UpdateVote UpId) UpdateVote m
+    , postSscCommitment :: KeyVal (Tagged MCCommitment StakeholderId) MCCommitment m
+    , postSscOpening    :: KeyVal (Tagged MCOpening StakeholderId) MCOpening m
+    , postSscShares     :: KeyVal (Tagged MCShares StakeholderId) MCShares m
+    , postSscVssCert    :: KeyVal (Tagged MCVssCertificate StakeholderId) MCVssCertificate m
+
+      -- Give a heavy delegation certificate. Returns False if something
+      -- went wrong.
+      --
+      -- NB light delegation is apparently disabled in master.
+    , postPskHeavy      :: ProxySKHeavy -> m Bool
+
+      -- Recovery mode related stuff.
+      -- TODO get rid of this eventually.
+    , recoveryInProgress :: m Bool
+
+      -- TBD does this need to be in the monad? It's not constant?
```

Comment:
Looks like it is constant.

Reply:
Right I guess it's monadic in existing code just because it's produced from a reader/lens combo.